File: 139. Word Break/solution.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        # edge case
        if not s or not wordDict:
            return False
        
        # create word set
        words = set(wordDict)
        
        # basic dfs
        visited = [False]*len(s)
        return self.dfs(words, visited, s, 0)
    
    def dfs(self, words, visited, string, start):
        if visited[start]:
            return False
        
        visited[start] = True
        
        for end in range(start, len(string)):
            if string[start:end+1] in words:
                if end == len(string) - 1:
                    return True
                else:
                    res = self.dfs(words, visited, string, end+1)
                    if res:
                        return True
                    
        return False
            

File: 139. Word Break/solution.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        HashSet<String> set = new HashSet<String>(wordDict);
        boolean[] visited = new boolean[s.length()];
        for (int i = 0; i < s.length(); i++) {
            visited[i] = false;
        }
        return wordBreak(s, 0, set, visited);
    }
    
    public boolean wordBreak(String s, int start, HashSet<String> set, boolean[] visited) {
        int n = s.length();
        visited[start] = true;
        
        for (int i = start; i < n; i++) {
            if (set.contains(s.substring(start, i+1))) {
                if (i + 1 == n) {
                    return true;
                }
                
                if (!visited[i+1] && wordBreak(s, i + 1, set, visited) == true) {
                    return true;
                }
            }
        }
        return false;
    }
}
